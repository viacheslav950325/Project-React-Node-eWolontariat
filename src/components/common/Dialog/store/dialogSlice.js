import { createSlice } from '@reduxjs/toolkit';

export const FormType = Object.freeze({
  rejestracja: 'rejestracja',
  zalozonyProfil: 'zalozonyProfil',
  loginDialog: 'loginDialog',
  rejestracjaEmail: 'rejestracjaEmail',
});

const initialState = {
  isOpen: false,
  formType: undefined,
};

const slice = createSlice({
  name: 'dialog',
  initialState,
  reducers: {
    openDialog: (state, action) => {
      return { ...action.payload, isOpen: true };
    },
    closeDialog: (state) => {
      state.isOpen = false;
      state.formType = undefined;
    },
  },
});

export const { openDialog, closeDialog } = slice.actions;

export default slice.reducer;
