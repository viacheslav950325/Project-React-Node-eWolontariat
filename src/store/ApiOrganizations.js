export default class Api {
   
    static getData() {
        const uri = 'https://api.npoint.io/0dd81319d981020bb21f';
        return fetch(uri, {
            method: 'GET'
        });
    }
 }
 