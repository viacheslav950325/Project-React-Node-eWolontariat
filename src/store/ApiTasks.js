export default class Api {
   
    static getData() {
        const uri = 'https://api.npoint.io/8d11637b99831903048a';
        return fetch(uri, {
            method: 'GET'
        });
    }
 }
 