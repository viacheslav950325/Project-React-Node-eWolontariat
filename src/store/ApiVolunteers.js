export default class Api {
   
    static getData() {
        const uri = 'https://api.npoint.io/3f0f98a4cd7eca33c3e9';
        return fetch(uri, {
            method: 'GET'
        });
    }
 }