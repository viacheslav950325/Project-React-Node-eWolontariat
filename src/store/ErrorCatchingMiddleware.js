import { isRejectedWithValue } from '@reduxjs/toolkit';
import { toast } from 'react-toastify';

export const rtkQueryErrorLogger = (api) => (next) => (action) => {

  if (isRejectedWithValue(action)){
    toast.warn(action.payload.message);
  }
  else if (action.error){
    toast.error(action.error.message);
  }

  return next(action);
}