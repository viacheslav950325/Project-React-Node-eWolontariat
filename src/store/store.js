import { configureStore, getDefaultMiddleware} from "@reduxjs/toolkit";
import taskReducer from "./taskSlice"
import volunteerReducer from "./volunteerSlice";
import dialogReducer from "../components/common/Dialog/store/dialogSlice";
import systemReducer from './systemSlice';
import organizationReducer from './organizationSlice'
//import { composeWithDevTools } from "@reduxjs/toolkit/dist/devtoolsExtension";

//const composedEnhancer = composeWithDevTools(applyMiddleware(thunkMiddleware));

export const store = configureStore({
    middleware: getDefaultMiddleware({
        serializableCheck: false
      }),
    reducer:{
        volunteers: volunteerReducer,
        dialog: dialogReducer,
        system: systemReducer,
        task: taskReducer,
        organization: organizationReducer,
    },
    //composedEnhancer
});

