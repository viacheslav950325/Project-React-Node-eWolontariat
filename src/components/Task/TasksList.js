import { Box, CircularProgress, Typography, Divider } from "@material-ui/core";
import { useEffect, useState  } from "react";
import { useSelector, useDispatch } from "react-redux";
import { sortTasks,  selectAllTasks } from "../../store/taskSlice";
import CustomButton from "../../theme/CustomButton";
import { fetchTasks } from "../../store/fetchTasks";
import getTasksCards from "./getTasksCards";
import SearchInputTasks from "../TasksPage/SearchInputTasks";
import ChooseCat from "../ChooseCat";
import setTasksRatingButtons from "../TasksPage/setTasksRatingButtons";

const TasksList = ({startSlice,endSlice}) => {
  const dispatch = useDispatch();
  const tasksList = useSelector(selectAllTasks);
  const tasksStatus = useSelector(state => state.task.status);
  const error = useSelector(state => state.task.error);
  const [filteredTasks, setTasks] = useState([]);
  const [isFilterTasks, setFilterTasks] = useState(false);

  let orderedTasks =[];

  useEffect(() => {
    if (tasksStatus === 'idle') {
      dispatch(fetchTasks())
    }
  }, [tasksStatus, dispatch]);

  let content, searchInput;

  if (tasksStatus === 'loading...') {
       content = (
        <Box style={{color: 'red'}} padding={2} align={"center"}>
            <CircularProgress style={{margin: "2rem"}} align={"center"} color={"secondary"}/>
        </Box>
      );
  } else if (tasksStatus === 'succeeded (:') {
      orderedTasks = dispatch(sortTasks(tasksList.tasks)).payload;
      searchInput = <SearchInputTasks/>;
      
  } else if (tasksStatus === 'failed :(') {
       content = <Box style={{color: 'red'}} padding={2} align={"center"}>ERROR: {error}</Box>;
  }  
  
  function getFilteredTextFromButton(text) {
    return (
        tasksList.tasks?.filter(element => 
              element.categories.includes(text))
          )
    };

    let thePopularCategoriesButtons = setTasksRatingButtons(orderedTasks).slice(0,4);

    return (
      <Box>
        <Box display={"flex"} justifyContent={"center"} style={{margin: "4rem 0 1rem"}}>
              {searchInput}
        </Box>
        <Box id="filtering-buttons" display={"flex"} justifyContent={'center'}  gridColumnGap={"2rem"} padding={"2rem 0"} margin={"1rem"} alignItems={"center"} flexWrap={"wrap"}>
                    <Typography variant="subtitle2" align={"left"} style={{marginTop: "1rem"}}>Najpopularniejsze <br/>kategorie:</Typography> 
                    <CustomButton 
                        variant="outlined" 
                        style={{marginTop: "1rem"}}
                        color={"primary"}  
                        onClick={()=>{
                          setFilterTasks(false);
                        }}        
                    >
                    Wszystkie</CustomButton>
                {thePopularCategoriesButtons.map((category,id)=>(
                    <CustomButton 
                        key={`item-${id}`} 
                        id= {category.value}
                        variant="contained" 
                        style={{marginTop: "1rem"}}
                        color={category.buttonColor}
                        startIcon={category.icon}
                        onClick={()=>{
                            setTasks(getFilteredTextFromButton(category.value));
                            setFilterTasks(true);
                        }}                           
                    >
                        <Divider orientation="vertical" flexItem style={{backgroundColor: "#eee", marginRight:"10px"}} /> {category.value}
                    </CustomButton>
                    
                ))}
            </Box>
            <Box display={"flex"} justifyContent={"space-evenly"} alignItems={"center"} flexWrap={"wrap"}>
              <ChooseCat/>
            </Box>
              {content}
              <Box display={'flex'} flexDirection={"row"} flexWrap={"wrap"} padding={'0 4rem 4rem 4rem'} justifyContent={'center'}>
                  {getTasksCards(isFilterTasks, orderedTasks, filteredTasks, startSlice, endSlice)}
              </Box>
         </Box>
    )
}

export default TasksList;
