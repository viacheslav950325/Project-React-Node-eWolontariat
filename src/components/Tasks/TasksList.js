import { Box, CircularProgress, Typography, Divider } from "@material-ui/core";
import { useEffect, useState  } from "react";
import { useSelector, useDispatch } from "react-redux";
import { sortTasks,  selectAllTasks  } from "../../store/taskSlice";
import TasksCard from "../TasksPage/TasksCard";
import Categories from "../../assets/data/Categories";
import setCategoryIcon from "../setCategoryIcon";
import CustomButton from "../CustomButton";
import { fetchTasks } from "../../store/fetchTasks";

const TasksList = () => {
  const dispatch = useDispatch();
  const tasksList = useSelector(selectAllTasks);
  const tasksStatus = useSelector(state => state.status);
  const error = useSelector(state => state.error);
  const [filteredTasks, setTasks] = useState([]);
  let orderedTasks =[];
  console.log(tasksList)

  useEffect(() => {
    if (tasksStatus === 'idle') {
      dispatch(fetchTasks())
    }
  }, [tasksStatus, dispatch]);

  let content;

  if (tasksStatus === 'loading...') {
    content = 
        (<CircularProgress style={{margin: "2rem"}} align={"center"} color={"secondary"}/>)
    
  } else if (tasksStatus === 'succeeded (:') {
    orderedTasks = (dispatch(sortTasks(tasksList.tasks))).payload.slice(0,6);
     
    content = orderedTasks.map((task,id) =>{
       return <TasksCard key={id} task={task} id={task.id}/>
    });
    
  } else if (tasksStatus === 'failed') {
    content = (
        <div style={{color: 'red'}}>ERROR: {error}</div>
    )
  }  

  console.log(tasksList)

  const thePopularCategories = (Categories.map(category=>{
    let rate=0;
    orderedTasks.tasks?.map(task=>{
        (task.categories).forEach(taskCategory=>{
            if(taskCategory===category.value){
                rate++;
            }
        });
        return rate;
    });
    category.rate=rate;
    category.buttonColor = setCategoryIcon(category.value)[1];
    category.icon = setCategoryIcon(category.value)[0];
    return category
    })).sort(compare).slice(0,4);

    function getFilteredTextFromButton(text) {
        return (
          orderedTasks?.filter(element => 
                element.categories.includes(text))
            )
      };

    return (
      <>
        <Box id="filtering-buttons" display={"flex"} justifyContent={'space-evenly'}  gridColumnGap={"2rem"} padding={"2rem 0"} margin={"1rem 3rem 0 3rem"} alignItems={"flex-end"}>
               <Box justifyContent={"flex-start"} flexGrow={"1"} >
                    <Typography variant="body2" align={"left"}>Najpopularniejsze kategorie:</Typography>
               </Box>
                    <CustomButton 
                        variant="outlined" 
                        color={"primary"}  
                        onClick={()=>{
                            setTasks(tasksList);
                        }}        
                    >
                    Wszystkie</CustomButton>

                {thePopularCategories.map((category,id)=>(
                    
                    <CustomButton 
                        key={`item-${id}`} 
                        id= {category.value}
                        variant="contained" 
                        color={category.buttonColor}
                        startIcon={category.icon}
                        onClick={()=>{
                            const filteredTasks = getFilteredTextFromButton(category.value);
                            setTasks(filteredTasks);
                            console.log(filteredTasks);
                        }}                           
                    >
                        <Divider orientation="vertical" flexItem style={{backgroundColor: "#eee", marginRight:"10px"}} /> {category.value}
                    </CustomButton>
                ))}
            </Box>
            <Box display={'flex'} flexDirection={"row"} flexWrap={"wrap"} padding={'0 4rem 4rem 4rem'} justifyContent={'center'}>
            {content}
            </Box>
         </>
    )
}

function compare( a, b ) {
    if ( a.rate > b.rate  ){
      return -1;
    }
    if ( a.rate < b.rate ){
      return 1;
    }
    return 0;
  };

export default TasksList;