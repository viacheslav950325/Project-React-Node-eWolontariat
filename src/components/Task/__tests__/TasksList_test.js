import toJson from "enzyme-to-json";
import { Provider } from "react-redux";
import { store } from "../../../store/store";
import TasksList from "../TasksList";

const tasksListWrapper = () => {
  return (
    <Provider store={store}> 
      <TasksList /> 
    </Provider>
  )
}

describe("TasksList",()=>{

    it("Should reender TasksList correctly", ()=>{
      expect(toJson(tasksListWrapper)).toMatchSnapshot();
      });

    it("Should filter tasks list correctly", ()=>{
     
      //expect(tasksListWrapper.instance().getFilteredTextFromButton([])).toEqual([]);
    });

    it("Should get tasks list correctly from Api", () =>{
        const mockFetch = jest.fn()
        .mockReturnValue(Promise.resolve({
          json: () => Promise.resolve({
            tasks: [
              {id:1, title: "Black Widow"},
              {id:3, title: "Kapitan Marvel"},
              {id:4, title: "Iron Man"},
            ]
          })
        }));
        expect.assertions(1);

        //dopisać jeszcze wywołanie fetcha 
        
          expect(mockFetch.mock.calls.length).toBe(1);
          expect(mockFetch).toBeCalledWith('https://api.npoint.io/8d11637b99831903048a');
       
    });
});