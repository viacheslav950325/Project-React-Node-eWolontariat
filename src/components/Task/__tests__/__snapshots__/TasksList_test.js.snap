// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksList Should reender TasksList correctly 1`] = `undefined`;
